File: ios/Classes/Services/PaymentServiceImpl.swift
//
//  PaymentServiceImpl.swift
//  Pods
//
//  Created by Angelo Cassano on 20/05/25.
//
import XPaySDK
import UIKit

public class PaymentServiceImpl: NSObject, PaymentApi {

    func pay(request: PaymentRequest, completion: @escaping (Result<PaymentResult, Error>) -> Void) {
        do {
            let xPay = try XPay(secretKey: request.secretKey)

            if let domain = request.domain, !domain.isEmpty {
                xPay._FrontOffice.setDomain(newUrl: domain)
            }

            xPay._FrontOffice.SelectedEnvironment = (request.environment == .production) ?
                EnvironmentUtils.Environment.prod :
                EnvironmentUtils.Environment.test

            let xPayRequest = ApiFrontOfficeQPRequest(
                alias: request.alias,
                codTrans: request.codTrans,
                currency: request.currency,
                amount: Int(request.amount)
            )

            guard let rootViewController = Self.findRootViewController() else {
                return completion(.failure(PigeonError(
                    code: "missing_rootVC",
                    message: "No root view controller available",
                    details: nil
                )))
            }

            xPay._FrontOffice.paga(
                xPayRequest,
                navigation: true,
                parentController: rootViewController
            ) { response in
                if response.IsValid {
                    completion(.success(response.IsCanceled ? .canceled : .completed))
                } else {
                    completion(.failure(PigeonError(
                        code: "payment_failed",
                        message: "An error has occurred during payment",
                        details: nil
                    )))
                }
            }

        } catch {
            completion(.failure(PigeonError(
                code: "exception",
                message: "A generic error has occurred",
                details: error.localizedDescription
            )))
        }
    }

    private static func findRootViewController() -> UIViewController? {
        UIApplication.shared.connectedScenes
            .compactMap { $0 as? UIWindowScene }
            .first(where: { $0.activationState == .foregroundActive })?
            .windows
            .first(where: { $0.isKeyWindow })?
            .rootViewController
    }
}

File: android/src/main/kotlin/it/angelocassano/flutter_nexi/services/payment/PaymentServiceImpl.kt
package it.angelocassano.flutter_nexi.services.payment

import android.app.Activity
import it.nexi.xpay.CallBacks.FrontOfficeCallbackQP
import it.nexi.xpay.Models.WebApi.Errors.ApiErrorResponse
import it.nexi.xpay.Models.WebApi.Requests.FrontOffice.ApiFrontOfficeQPRequest
import it.nexi.xpay.Models.WebApi.Responses.FrontOffice.ApiFrontOfficeQPResponse
import it.nexi.xpay.Utils.EnvironmentUtils
import it.nexi.xpay.XPay

class PaymentServiceImpl : PaymentApi {

    var activity: Activity? = null

    override fun pay(request: PaymentRequest, callback: (Result<PaymentResult>) -> Unit) {
        val activity = this.activity
        if (activity == null) {
            callback(Result.failure(IllegalStateException("Activity is null")))
            return
        }

        try {
            val xPay = XPay(activity, request.secretKey)

            xPay.FrontOffice.setEnvironment(
                if (request.environment == Environment.PRODUCTION)
                    EnvironmentUtils.Environment.PROD
                else
                    EnvironmentUtils.Environment.TEST
            )

            request.domain?.takeIf { it.isNotBlank() }?.let {
                xPay.FrontOffice.setDomain(it)
            }

            val xPayRequest = ApiFrontOfficeQPRequest(
                request.alias,
                request.codTrans,
                request.currency,
                request.amount
            )

            xPay.FrontOffice.paga(xPayRequest, true, object : FrontOfficeCallbackQP {

                override fun onConfirm(response: ApiFrontOfficeQPResponse) {
                    if (response.isValid) {
                        callback(Result.success(PaymentResult.COMPLETED))
                    } else {
                        val message = response.error?.message ?: "Unknown payment error"
                        callback(Result.failure(RuntimeException(message)))
                    }
                }

                override fun onCancel(response: ApiFrontOfficeQPResponse) {
                    callback(Result.success(PaymentResult.CANCELED))
                }

                override fun onError(error: ApiErrorResponse) {
                    callback(Result.failure(RuntimeException(error.error.message)))
                }
            })

        } catch (e: Exception) {
            callback(Result.failure(e))
        }
    }
}
